Stacktask

Iweinrazvadovskaya opened the pull request:
Hello. Check my solution, please =) Should I rewrite my code without using .min() function

I’m not sure that is a good way to find out minimal item in another function. Should function do more than one thing (stackStatistic class push and pop functions)?

dersim-davaod commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+}
+
+class StackStatistic: Stack {
```
```suggestion
class StackStatistics: Stack {
```

Please read the task one more time.

Iweinrazvadovskaya replied:
ok)

dersim-davaod commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,96 @@
+        if stackArray.isEmpty {
+            minimalItem = element
+        }
+        if element < minimalItem! {
```
Please update your code to avoid using `!` operator.

Iweinrazvadovskaya replied:
can I use `!` in cases like after update ?

dersim-davaod commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+    
+    // Function finds minimal element in stack, returns the top element of stack and remove this element.
+    override func pop() -> Int? {
```
Please evaluate the performance of the `pop()` using big-O notation. Leave the comment.

Iweinrazvadovskaya replied:
O(1) after updating.

Iweinrazvadovskaya commented on README.md:
```diff
@@ -0,0 +1,27 @@
+--------------------------------------------------------------------------------------------------------------
+# Stack Task
+
+### A stack is a container of objects that are inserted and removed according to the last-in first-out (LIFO) principle. In the pushdown stacks only two operations are allowed: 
```
got it )

Iweinrazvadovskaya commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,96 @@
+// Protocol contains methods for working with Stack class
```
Yes) I agree

Iweinrazvadovskaya commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,96 @@
+        if stackArray.isEmpty {
+            return nil
+        } else {
+            let lastStackItem = stackArray[stackArray.count-1]
```
got it

Iweinrazvadovskaya commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,96 @@
+            return nil
+        } else {
+            let lastStackItem = stackArray[stackArray.count-1]
+            stackArray.remove(at: stackArray.count-1)
```
Alright

Iweinrazvadovskaya commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,96 @@
+}
+
+class Stack: StackOperations {
+    fileprivate var stackArray: [Int] = []
```
got it!

dersim-davaod commented on README.md:
```diff
@@ -0,0 +1,27 @@
+O(N)
+
+.mode
+O(N^2)
```
Please see how it's rendered in browser (you can use `Display the rich diff button`:
<img width="240" alt="Screen Shot 2020-02-19 at 12 24 10 PM" src="https://user-images.githubusercontent.com/6073510/74820253-c225df00-5312-11ea-82f2-46d52b467df0.png">

Iweinrazvadovskaya replied:
ok.

dersim-davaod replied:
Did you use the button?

Iweinrazvadovskaya replied:
Now yes

dersim-davaod commented on README.md:
```diff
@@ -0,0 +1,27 @@
+# First Task - Big O
+
+.generateRandomNumber
```
Please add a more clear description. What is `generateRandomNumber`? What is `makeRandomArray`?
Why you have chosen them? For now it seems like just a list of some names and complexities (time/space, I don't know).

Iweinrazvadovskaya replied:
got it)

dersim-davaod commented on README.md:
```diff
@@ -0,0 +1,27 @@
+
+.mode
+O(N^2)
+--------------------------------------------------------------------------------------------------------------
```
Is that a snake? :)

Iweinrazvadovskaya replied:
how did you guess?)

Iweinrazvadovskaya commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,74 @@
+stackStatistics.push(element: 5)
+stackStatistics.push(element: 3)
+stackStatistics.pop()
+print ("Minimum  element in array (if answer = nil, array is empty ): \(String(describing: stackStatistics.minimumElement()))")
```
ok

Iweinrazvadovskaya commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,74 @@
+    
+    // Function returns the top element of the stack, but does not remove it.
+    func top() -> Int? {
+        stackArray.isEmpty ? nil : stackArray.last
```
of course! It's a failed mistake(

Iweinrazvadovskaya commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,74 @@
+
+    // Function adds element on the top of the steck and finds minimal element
+    override func push(element: Int) {
+        if  minimalElements.isEmpty || element <= minimalElements.last!{
```
I will

dersim-davaod commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,75 @@
+
+    // Function returns minimum element in the stack.
+    func minimumElement() -> Int? {
+        minimalElements.isEmpty ? nil : minimalElements.last
```
Please see https://github.com/Iweinrazvadovskaya/CourseHomeTasks/pull/3#discussion_r380488439

Iweinrazvadovskaya replied:
thank u )

dersim-davaod commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,73 @@
+        super.push(element: element)
+    }
+    
+    // Function finds minimal element in stack, returns the top element of stack and remove this element.
```
1) In fact, your function do not find the min element.

2) The comment says:
```
... finds minimal element, returns the top element of stack and remove this element.
```

What does that mean `this element`? Does it remove the min element or the top element?

Iweinrazvadovskaya replied:
I will rewrite it.

dersim-davaod commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,73 @@
+        return droppedElement
+    }
+
+    // Function returns minimum element in the stack.
```
Could you please explain, why do you think this comment should be here? Just want to understand the reason and your opinion.

Iweinrazvadovskaya replied:
To explain what this function does. Something is wrong?

dersim-davaod replied:
Ok for now.
But in general, such comments are useless. It's clear from the code what the function does.

Usually, you have to use doc comments to explain an important implementation details or API.

Iweinrazvadovskaya replied:
ok

dersim-davaod commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+    
+    // Function returns the top element of the stack, but does not remove it.
```
You left a comment about `nil` for the function `pop()`:
```
returns nil when the invoking stack is empty.
```

Why did you decide to avoid it here?

Iweinrazvadovskaya replied:
I will rewrite it

dersim-davaod commented on StackTask.playground/Contents.swift:
```diff
@@ -0,0 +1,73 @@
+        stackArray.append(element)
+    }
+    
+    // Function removes and returns the top element of the stack. returns nil when the invoking stack is empty.
```
```suggestion
    // Function removes and returns the top element of the stack. Returns nil when the invoking stack is empty.
```

Iweinrazvadovskaya replied:
ok

Iweinrazvadovskaya commented on README.md:
```diff
@@ -0,0 +1,46 @@
+
+Overriding base function: uses the base implementation and looks for minimum element (if was remover minimal element)
+
+> Finding the minimum element is realized by adding the smallest or the same element to the array of minimal elements `minimalElements`.
```
![image](https://user-images.githubusercontent.com/43065967/75116450-9a6ba980-5679-11ea-9f57-2aa5945c632b.png)
why not& :)

Iweinrazvadovskaya replied:
![image](https://user-images.githubusercontent.com/43065967/75116477-dbfc5480-5679-11ea-8e75-47f8756fd467.png)
why not? :)

dersim-davaod replied:
Hm, technically, yes.
Let me check it with native speaker. AFAIK in contemporary English the word `realise` is a synonym to `understand`.

Try to search the word in context:
`"function is implemented"` ~ 15 500 000 results according to Google.
`"function is realized"` ~ 2 640 000 results.

For now, I won't treat this as an error.

Iweinrazvadovskaya replied:
ok )

Iweinrazvadovskaya commented on README.md:
```diff
@@ -0,0 +1,46 @@
+
+Overriding base function: uses the base implementation and looks for minimum element (if was remover minimal element)
+
+> Finding the minimum element is realized by adding the smallest or the same element to the array of minimal elements `minimalElements`.
```
I saw, that people sometimes use quotes for the selection part of the text. But I will remove it, as u want :)

dersim-davaod replied:
Why do you think I want you to remove it?

dersim-davaod replied:
Usually, people do NOT use quotes just for "selection". People use it to *quote* something (surprise!).

dersim-davaod replied:
Your quotation made people thinking: It is a quote, isn't it? Who owns the quote? Why it's important to quote it? etc.

Iweinrazvadovskaya replied:
Thanks, got it. U r right

Iweinrazvadovskaya commented on .gitignore:
```diff
@@ -1,6 +1,3 @@
-# Created by https://www.gitignore.io/api/xcode,macos,swift
```
ok, I will return the link

Iweinrazvadovskaya commented on .gitignore:
```diff
@@ -13,23 +10,6 @@ Icon
 ._*
 
-# Files that might appear in the root of a volume
-.DocumentRevisions-V100
```
I understand your message. I will observe in the future

Iweinrazvadovskaya commented:
@dersim-davaod  
I deleted this file from this branch. It was a mistake, thank you for correcting

> Why do you think file CourseHomeTasks/ArrayOperations.playground/playground.xcworkspace/xcuserdata/jane.xcuserdatad/UserInterfaceState.xcuserstate should be here?
I fixed it.
> Please open the PR in Github and take a look at the bottom of Conversation tab. The GitHub says:
> This branch has conflicts that must be resolved
